Backport a bristuff deadlock fix from changes made to chan_dahdi by Tzafrir.
(commit ccd11da0599c190f5b678aed3f164579ca873c71 on Xorcom's tree)

 -- Faidon Liambotis <paravoid@debian.org>

--- a/channels/chan_zap.c
+++ b/channels/chan_zap.c
@@ -8851,7 +8851,7 @@ static void *pri_dchannel(void *vpri)
 	int haveidles;
 	int activeidles;
 	int nextidle = -1;
-	struct ast_channel *c;
+	struct ast_channel *c = NULL;
 	struct timeval tv, lowest, *next;
 	struct timeval lastidle = { 0, 0 };
 	int doidling=0;
@@ -9590,6 +9590,7 @@ static void *pri_dchannel(void *vpri)
 								snprintf(calledtonstr, sizeof(calledtonstr)-1, "%d", e->ring.calledplan);
 								pbx_builtin_setvar_helper(c, "CALLEDTON", calledtonstr);
 
+								ast_mutex_lock(&c->lock);
 								ast_mutex_lock(&pri->pvts[chanpos]->lock);
 								ast_mutex_lock(&pri->lock);
 
@@ -9637,6 +9638,8 @@ static void *pri_dchannel(void *vpri)
 					if (crv)
 						ast_mutex_unlock(&crv->lock);
 					ast_mutex_unlock(&pri->pvts[chanpos]->lock);
+					if (c)
+						ast_mutex_unlock(&c->lock);
 				} else {
 					if (e->ring.flexible)
 						pri_hangup(pri->pri, e->ring.call, PRI_CAUSE_NORMAL_CIRCUIT_CONGESTION, -1);
